server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotsServiceIT.java

```diff
@@ -111,6 +112,42 @@ public void testSnapshotDeletionFailureShouldBeLogged() throws Exception {
         }
     }
 
+    public void testDeleteSnapshotWhenNotWaitingForCompletion() throws Exception {
+        createIndexWithRandomDocs("test-index", randomIntBetween(1, 5));
+        createRepository("test-repo", "mock");
+        createSnapshot("test-repo", "test-snapshot", List.of("test-index"));
+        MockRepository repository = getRepositoryOnMaster("test-repo");
+        PlainActionFuture<AcknowledgedResponse> listener = new PlainActionFuture<>();
+        repository.blockOnDataFiles();
+        try {
+            clusterAdmin().prepareDeleteSnapshot("test-repo", "test-snapshot").setWaitForCompletion(false).execute(listener);
+            listener.get(5, TimeUnit.SECONDS);
+            assertNotNull(getSnapshot("test-repo", "test-snapshot"));
+        } finally {
+            repository.unblock();
+        }
+        assertBusy(() -> assertThrows(SnapshotMissingException.class, () -> getSnapshot("test-repo", "test-snapshot")));
+    }
+
+    public void testDeleteSnapshotWhenWaitingForCompletion() throws Exception {
+        createIndexWithRandomDocs("test-index", randomIntBetween(1, 5));
+        createRepository("test-repo", "mock");
+        createSnapshot("test-repo", "test-snapshot", List.of("test-index"));
+        MockRepository repository = getRepositoryOnMaster("test-repo");
+        PlainActionFuture<AcknowledgedResponse> listener = new PlainActionFuture<>();
+        repository.blockOnDataFiles();
+        try {
+            clusterAdmin().prepareDeleteSnapshot("test-repo", "test-snapshot").setWaitForCompletion(true).execute(listener);
+            // The listener won't be resolved, and snapshot won't be deleted until we remove the block
+            assertFalse(listener.isDone());
+            assertNotNull(getSnapshot("test-repo", "test-snapshot"));
+        } finally {
+            repository.unblock();
+        }
+        listener.get(5, TimeUnit.SECONDS);
```

Comment:
5 seconds is a little short for some CI runners - I suggest `safeGet(listener)` here.